### DISCUSSION TITLE
Load testing with Playwright using multiple concurrent users

### DISCUSSION THREAD

**Discussion by ksatishcrm** [artilleryio/artillery#1643] [Artillery]
Using Playwright, we are testing our test cases for Dynamics 365 and with single user we were able to run our test cases successfully.
But we are struck with load testing, when using multiple concurrent users for running our test cases.

We followed below documentation for multiple users authentication.
https://playwright.dev/docs/auth#avoiding-multiple-sessions-per-account-at-a-time

Can anyone share insights on how to implement loading testing with multi concurrent users.
Can you please let us know if this is possible through playwright engine artillery to achieve load testing.

**Comment by hassy** [comment]
hi @ksatishcrm 👋 Artillery launches a separate browser / Playwright instance for each virtual user.

**Comment by ksatishcrm** [comment]
Thanks Hassy. Can you please share any reference/code snippet for multi user authentication for load test using Playwright engine Artillery.

**Reply by hassy** [reply]
We don't have an example for that use case specifically, but there's an end-to-end example of load testing with Playwright here: https://github.com/artilleryio/artillery/tree/master/examples/browser-load-testing-playwright

It can serve as a starting point for you to adapt your Playwright code for load testing.

Feel free to share any code snippets if you require some help as well!

**Comment by ksatishcrm** [comment]
@hassy, Does Artillery support only API load testing or UI load test as well?

**Reply by hassy** [reply]
It can be used for both @ksatishcrm

**Comment by ksatishcrm** [comment]
Thanks @hassy for letting us know that UI load test can be handled by Artillery. 
The example which you shared look to be for API testing.

In our case we are using Dynamics 365 and we have few test cases created and want to capture Load test performance.
We will load Dynamics 365 and traverse to one of the entities from UI, which will load the records from the view of an entity requested. 
And this test case is generated using CodeGen.

Is it possible to use the test case that is generated using CodeGen and use it for our UI load testing.
Can you please let us know how to pass test case in artillery and capture UI Load test performance.
And also wanted to know how to create YAML for the test case generated from CodeGen.

Can you share if you have any code snippet for the same.

**Comment by ksatishcrm** [comment]
We are planning to create load testing for our application using Artillery with playwright with multi users.
In below scenario we are passing 2 users for testing, and we have created our test cases under LoadTest.js.
We observed that our test cases are not running but every second  **function: "runTest"** is getting hit.

Our test cases are not running inside the function: "runTest" can you please let us know how to configure the phases for our scenario.
We went through the documentation(https://www.artillery.io/docs/guides/guides/test-script-reference) but unable to understand the configuration for load testing.

Can you please share some insights how to use phases.
Can you please also let us know how to store the authentication state and pass the token to the test cases.


config:

  #tls: {
  #  "rejectUnauthorized": false
  #}
  phases:
    - duration: 30
      arrivalRate: 2
      rampTo: 1
  payload:
    path: "./users.csv"
    fields:
      - "username"
      - "password"
    order: sequence    
  processor: "./LoadTest.js"

scenarios:
  - flow:
    - function: "runTest"
    - log: "Current username is set to: {{ $username }}"

**Reply by hassy** [reply]
hi @ksatishcrm - can you describe the load profile you want to configure?

The phase definition that you shared:

```yaml
- duration: 30
  arrivalRate: 2
  rampTo: 1
```

Does this:
- Start by creating two VUs every second (that's the meaning of "arrival rate" in Artillery)
- Gradually decrease the number of arrivals to 1 per second (that's the meaning of "ramp" in Artillery)
- This will happen over the period of 30 seconds.
